src/System.Diagnostics.Tracing/tests/BasicEventSourceTest/Harness/Listeners.cs

```diff
@@ -0,0 +1,433 @@
+    public class EtwListener : Listener
+    {
+        internal static void InsureStopped()
+        {
+            using (var session = new TraceEventSession("EventSourceTestSession", "EventSourceTestData.etl"))
+                session.Stop();
+        }
+
+        public EtwListener(string dataFileName = "EventSourceTestData.etl", string sessionName = "EventSourceTestSession")
+        {
+            m_dataFileName = dataFileName;
+
+            // Today you have to be Admin to turn on ETW events (anyone can write ETW events).   
+            if (TraceEventSession.IsElevated() != true)
+            {
+                Assert.Inconclusive("Need to be admin elevated to run this test.");
+                throw new ApplicationException("Need to be elevated to run. ");
+            }
+
+            if (dataFileName == null)
+            {
+                Console.WriteLine("Creating a real time session " + sessionName);
+
+                Task.Factory.StartNew(delegate()
+                {
+                    var session = new TraceEventSession(sessionName, dataFileName);
+                    session.Source.AllEvents += OnEventHelper;
+                    Console.WriteLine("Listening for real time events");
+                    m_session = session;    // Indicate that we are alive.  
+                    m_session.Source.Process();
+                    Console.WriteLine("Real time listening stopping.");
+                });
+
+                while (m_session != null)   // Wait for real time thread to wake up. 
+                    Thread.Sleep(1);
+            }
+            else
+            {
+                // Normalize to a full path name.  
+                dataFileName = Path.GetFullPath(dataFileName);
+                Console.WriteLine("Creating ETW data file " + Path.GetFullPath(dataFileName));
+                m_session = new TraceEventSession(sessionName, dataFileName);
+            }
+        }
+        public override void EventSourceCommand(string eventSourceName, EventCommand command, FilteringOptions options=null)
+        {
+            if (command == EventCommand.Enable)
+            {
+                if (options == null)
+                    options = new FilteringOptions();
+
+                m_session.EnableProvider(eventSourceName, (TraceEventLevel)options.Level, (ulong)options.Keywords,
+                    new TraceEventProviderOptions() { Arguments = options.Args });
+            }
+            else if (command == EventCommand.Disable)
+            {
+                m_session.DisableProvider(TraceEventProviders.GetEventSourceGuidFromName(eventSourceName));
+            }
+            else
+                throw new NotImplementedException();
+            Thread.Sleep(200);          // Calls are async, give them time to work.  
+        }
+        public override void Dispose()
+        {
+            m_session.Flush();
+            Thread.Sleep(1010);      // Let it drain.  TODO not clear this is needed.  
```

Comment:
Any TODOs left in the code when this is going to be merged should have one or more issues opened and the issue number included with the TODO.